text/3637-guard-patterns.md

```diff
@@ -0,0 +1,274 @@
+
+```rust
+pattern if expr
+```
+
+This is a **guard pattern**. It matches a value if `pattern` (the pattern it wraps) matches that value, _and_ `expr` evaluates to `true`. Like in match arm guards, `expr` can use values bound in `pattern`.
+
+## For New Users
+
+For new users, guard patterns are better explained without reference to match arm guards. Instead, they can be explained by similar examples to the ones currently used for match arm guards, followed by an example showing that they can be nested within other patterns and used outside of match arms.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Supersession of Match Arm Guards
+
+Rather than being parsed as part of the match expression, guards in match arms will instead be parsed as a guard pattern. For this reason, the `if` pattern operator must have lower precedence than all other pattern operators.
+
+That is,
+
+```rs
+// Let <=> denote equivalence of patterns.
+
+x @ A(..) if pred       <=> (x @ A(..)) if pred
+&A(..) if pred          <=> (&A(..)) if pred
+A(..) | B(..) if pred   <=> (A(..) | B(..)) if pred
+```
+
+## Precedence Relative to `|` 
+
+Consider the following match expression:
+
+```rust
+match foo {
+    A | B if c | d => {},
+}
+```
+
+This match arm is currently parsed as `(A | B) if (c | d)`, with the first `|` being the or-operator on patterns and the second being the bitwise OR operator on expressions. Therefore, to maintain backwards compatability, `if` must have lower precedence than `|` on both sides (or equivalently, for both meanings of `|`). For that reason, guard patterns nested within or-patterns must be explicitly parenthesized:
+
+```rust
+// This is not an or-pattern of guards:
+    a if b | c if d
+<=> (a if (b | c)) if d
+
+// Instead, write
+(a if b) | (c if d)
+```
+
+## In Assignment-Like Contexts
+
+There's an ambiguity between `=` used as the assignment operator within the guard
+and used outside to indicate assignment to the pattern (e.g. in `if let`)
+Therefore guard patterns appearing at the top level in those places must also be parenthesized:
+
+```rust
+// Not allowed:
+let x if guard(x) = foo() {}
+if let x if guard(x) = foo() {}
+while let x if guard(x) = foo() {}   
+
+// allowed
+let (x if guard(x)) = foo() {} // Note that this would still error after parsing, since guard patterns are always refutable.
+if let (x if guard(x)) = foo() {}
+while let (x if guard(x)) = foo() {} 
+```
+
+Therefore the syntax for patterns becomes
+
+> **<sup>Syntax</sup>**\
+> _Pattern_ :\
+> &nbsp;&nbsp; &nbsp;&nbsp; _PatternNoTopGuard_\
+> &nbsp;&nbsp; | _GuardPattern_
+> 
+> _PatternNoTopGuard_ :\
+> &nbsp;&nbsp; &nbsp;&nbsp; `|`<sup>?</sup> _PatternNoTopAlt_  ( `|` _PatternNoTopAlt_ )<sup>\*</sup>
+
+With `if let` and `while let` expressions now using `PatternNoTopGuard`. `let` statements and function parameters can continue to use `PatternNoTopAlt`.
```

Comment:
A `for` expression also accepts a `Pattern`, but the pattern is required to be irrefutable. 

```rust
    for Ok(x) | Err(x) in vec![Ok(5), Err(6), Ok(7), Err(8)] {
        println!("{x}");
    }
```

I think this be changed to PatternNoTopGuard too?

| Expression | Irrefutable | Current grammar | This RFC |
|---|---|---|---|
| `let`(/`else`) | | PatternNoTopAlt | PatternNoTopAlt |
| Function & closure parameter | ✓ | PatternNoTopAlt | PatternNoTopAlt |
| `match` | | Pattern MatchArmGuard<sup>?</sup> | Pattern |
| `if`/`while let` | | Pattern | PatternNoTopGuard |
| `for` | ✓ | Pattern | PatternNoTopGuard |

Reply:
They could be, but wouldn't it be better to give an error saying the pattern must be refutable rather than a parse error?